src/sortTemplate.ts

```diff
@@ -0,0 +1,297 @@
+import { DeploymentTemplate } from "./DeploymentTemplate";
+import { ext } from './extensionVariables';
+import { IParameterDefinition } from './IParameterDefinition';
+import * as Json from "./JSON";
+import * as language from "./Language";
+import { UserFunctionDefinition } from './UserFunctionDefinition';
+import { UserFunctionNamespaceDefinition } from './UserFunctionNamespaceDefinition';
+import { IVariableDefinition } from './VariableDefinition';
+
+export enum SortType {
+    Resources,
+    Outputs,
+    Parameters,
+    Variables,
+    Functions,
+    TopLevel
+}
+
+export class SortQuickPickItem implements vscode.QuickPickItem {
+    public label: string;
+    public value: SortType;
+    public description: string;
+
+    constructor(label: string, value: SortType, description: string) {
+        this.label = label;
+        this.value = value;
+        this.description = description;
+    }
+}
+
+export function getQuickPickItems(): SortQuickPickItem[] {
+    let items: SortQuickPickItem[] = [];
+    items.push(new SortQuickPickItem("Functions", SortType.Functions, "Sort function namespaces and functions"));
+    items.push(new SortQuickPickItem("Outputs", SortType.Outputs, "Sort outputs"));
+    items.push(new SortQuickPickItem("Parameters", SortType.Parameters, "Sort parameters for the template"));
+    items.push(new SortQuickPickItem("Resources", SortType.Resources, "Sort resources based on the name including first level of child resources"));
+    items.push(new SortQuickPickItem("Variables", SortType.Variables, "Sort variables"));
+    items.push(new SortQuickPickItem("Top level", SortType.TopLevel, "Sort top level items based on recommended order (parameters, functions, variables, resources, outputs)"));
+    return items;
+}
+
+export async function sortTemplate(template: DeploymentTemplate | undefined, sortType: SortType): Promise<void> {
+    if (template === undefined) {
+        return;
+    }
+    ext.outputChannel.appendLine("Sorting template");
+    switch (sortType) {
+        case SortType.Functions:
+            await sortFunctions(template);
+            break;
+        case SortType.Outputs:
+            await sortOutputs(template);
+            break;
+        case SortType.Parameters:
+            await sortParameters(template);
+            break;
+        case SortType.Resources:
+            await sortResources(template);
+            break;
+        case SortType.Variables:
+            await sortVariables(template);
+            break;
+        case SortType.TopLevel:
+            await sortTopLevel(template);
+            break;
+        default:
+            vscode.window.showWarningMessage("Unknown sort type!");
+            return;
+
+    }
+    vscode.window.showInformationMessage("Done sorting template!");
+}
+
+async function sortOutputs(template: DeploymentTemplate): Promise<void> {
+    let rootValue = Json.asObjectValue(template.getJSONValueAtDocumentCharacterIndex(1));
```

Comment:
This assumes that the template starts with "{" on the first line.  If there's whitespace or comments at the top of the file, this returns undefined and sorting gets skipped.  You need to expose DeploymentTemplate's _topLevelValue property and use that (can be undefined if the file is malformed).

Reply:
Fixed.